src/main/java/seedu/address/model/task/Task.java

```diff
@@ -18,31 +20,65 @@
     private final Name name;
     private final Priority priority;
 
-    // Data fields
+    // Data fields'
     private final Description description;
     private final Done done;
     private final Set<Tag> tags = new HashSet<>();
+    private final Optional<Reminder> optionalReminder;
 
     /** Every field must be present and not null. */
+    public Task(Name name, Priority priority, Description description, Done done, Set<Tag> tags, Optional<Reminder> optionalReminder) {
+        requireAllNonNull(name, priority, description, tags);
+        this.name = name;
+        this.priority = priority;
+        this.description = description;
+        this.done = done;
+        this.tags.addAll(tags);
+        this.optionalReminder = optionalReminder;
+        triggerReminderIfPresent();
+    }
+
+    /** With done and no reminder */
     public Task(Name name, Priority priority, Description description, Done done, Set<Tag> tags) {
         requireAllNonNull(name, priority, description, tags);
         this.name = name;
         this.priority = priority;
         this.description = description;
         this.done = done;
         this.tags.addAll(tags);
+        this.optionalReminder = Optional.empty();
     }
 
     // without done provided
+    public Task(Name name, Priority priority, Description description, Set<Tag> tags, Optional<Reminder> optionalReminder) {
+        requireAllNonNull(name, priority, description, tags);
+        this.name = name;
+        this.priority = priority;
+        this.description = description;
+        this.done = new Done();
+        this.optionalReminder = optionalReminder;
+        this.tags.addAll(tags);
+        triggerReminderIfPresent();
+    }
+
+    // without Reminder or done provided
     public Task(Name name, Priority priority, Description description, Set<Tag> tags) {
         requireAllNonNull(name, priority, description, tags);
         this.name = name;
         this.priority = priority;
         this.description = description;
         this.done = new Done();
+        this.optionalReminder = Optional.empty();
         this.tags.addAll(tags);
     }
 
+    public void triggerReminderIfPresent() {
+        if (optionalReminder.isPresent()) {
+            Reminder reminder = optionalReminder.get();
+            MainWindow.triggerReminder(reminder, name.toString(), description.toString());
```

Comment:
This should be done in LogicManager:
1. to avoid cyclic dependencies -> MainWindow relies on Task, Task relies on MainWindow
2. This reminder is triggered everytime a Task is created, in our TaskList manager, Tasks are always recreated because we're using ReadOnlyTask., so even when it's not the reminder that was edited, a reminder will be triggered. Not sure if setting multiple identical timelines will cause any issues
3.  I don't think this covers cases where reminders are removed, we would need to stop the reminder on the MainWindow then.